### ISSUE TITLE
slack-notifications to multiple channels failing

### ISSUE THREAD

**Issue by chakrapani123** [ashwanthkumar/gocd-slack-build-notifier#62]
Hi,

i have a set-up where in all pipelines notifications of build be sent to dev-channel and deployment notifications sent to deploy-channel .

here is a snapshot of my config 

gocd.slack {

  login = “XXXX”

  password = “XXXXX”

  server-host = “XXXXXXX“

  api-server-host = “XXXXXX”

  webhookUrl = "https://hooks.slack.com/services/XXXXXXX”

  pipelines = [{
    name = “deploy-to-dev”
    stage = “deploy”
    state = "failed|passed"
    channel = “deployment-ops”
    webhookUrl = "https://hooks.slack.com/services/YYYYY”
  },
  {
    name = ".*"
    stage = ".*"
    state = "failed|passed"
    channel = “dev-channel”
  },
  {
    name = ".*"
    stage = ".*"
    state = "failed"
  }]
  # optional fields
  process-all-rules = "true"

}

unfortunately , notifications are not getting pushed to "deployment-ops" channel rule for "deploy-to-dev" pipeline but rest of the notifications to dev-channel are working. i have version 1.3.1 of plugin and on gocd version 16.1.

in other words , webhook url set globally seems to be working fine but not locally set within pipelines are working . i have 2 webhooks for 2 orgs in slack

**Comment by jdavisp3** [comment]
To use the `process-all-rules` option you will need to upgrade the plugin to at least `1.4.0-RC8`.

**Comment by chakrapani123** [comment]
tried with 1.4.0-RC8 still same issue. just to make sure , if  we have to enable notifications to be sent to multiple channels , we would be needing 2 webhooks and if webhook url is defined locally which would override global settings, notifications fail

**Comment by jdavisp3** [comment]
I only use one webhook URL and just override the channel everywhere. But multiple URLs should work too. Not sure what's going wrong with your installation, I am using `process-all-rules` in production now and sending the same event to multiple channels.

**Comment by ashwanthkumar** [comment]
@chakrapani123 Can you please share the plugin logs? It's generally located along with the go-server logs. It might have some answers.

**Comment by chakrapani123** [comment]
Hi,

here is the actual snippet just to have reference to actual log attached here

 pipelines = [{
    name = "xhsecure-service-deploy-to-qa"
    stage = "deploy-to-qa"
    state = "failed|passed"
    channel = "sanket-test"
    webhookUrl = "https://hooks.slack.com/services/XXXXXX”
  },
  {
    name = "._"
    stage = "._"
    state = "failed|passed"
    channel = "om-team-gocd"
  },
  {
    name = "._"
    stage = "._"
    state = "failed"
  }]
i have fixed my extra characters issue and it seems to be working now. However , its sending notifications for every stage instead of specific to deploy-to-qa stage. is that expected? or just taking regex of the actual pipeline name and sending out every stage notification?
[plugin-slack.notifier.txt](https://github.com/ashwanthkumar/gocd-slack-build-notifier/files/563255/plugin-slack.notifier.txt)

**Comment by ashwanthkumar** [comment]
That's probably because you've `process-all-rules = "true"` - if the first rule doesn't match the next rule for "all pipelines", "all stages" on passed/failed states send it to `#om-team-gocd` would match and probably why you're getting notifications.

**Comment by ashwanthkumar** [comment]
Was thinking if I commented too soon. Is that the problem you're facing? The notifications you're getting are on `#om-team-gocd` or `#sanket-test` ? 

If it's the former then it's probably because of the scenario I explained in the earlier comment. If it's the latter, then it's probably a bug in the plugin.

**Comment by chakrapani123** [comment]
It's the latter and notifications are sent to sanket-test for all xhs\* pipelines

**Comment by chakrapani123** [comment]
just a quick update. i have the same configuration as mentioned above and all the slack notifications go to sanket-test and nothing to om-team-gocd. in essence, plusin always sends to single channel irrespective of pipeline stage or pipeline. multi-channel doesnt seem to work and local variable always overrides global and thus pushing to latest rule than all the rules

**Comment by jdavisp3** [comment]
So I have a setup with GoCD 16.8 and 1.4.0-RC9 of the plugin with multiple channels, pipelines, etc. Here's a partial sample:

``` json
gocd : {
  slack : {
      "api-server-host" : "http://localhost:8153/",
      "channel" : "#gocd",
      "display-console-log-links" : false,
      "displayMaterialChanges" : false,
      "login" : "notification",
      "password" : "XXXX",
      "pipelines" : [
          {
              "channel" : "#build-minders",
              "name" : "website-(deploy|check)-(stage|prod)",
              "stage" : ".*",
              "state" : "all"
          },
          {
              "channel" : "#dev",
              "name" : "website-(deploy|check)-(dev|stage)",
              "stage" : ".*",
              "state" : "broken|cancelled|failed"
          }
      ],
      "process-all-rules" : true,
      "server-host" : "XXXX",
      "slackDisplayName" : "gocd",
      "webhookUrl" : "XXX"
  }
}
```

This is working fine for me, getting notifications in the right channels for the right pipelines, etc.

When I go to the plugins section on GoCD I see:

<img width="588" alt="screen shot 2016-11-02 at 12 05 07 pm" src="https://cloud.githubusercontent.com/assets/43582/19943449/aa910dac-a0f4-11e6-95fd-4adbefbd559f.png">

**Comment by chakrapani123** [comment]
thanks for sharing that.. i have a typical use-case where i have slack channels in 2 different org spaces in slack and thats the reason why incoming webhooks are prompting for 2 slack webhook urls instead of one and wehbook url defined locally overwrites global config or vice versa but doesnt work with 2 webhook urls and 2 channels

**Comment by chakrapani123** [comment]
here is my config

``` json
gocd.slack {

  login = “XXXX”

  password = “XXXX”
  server-host = “XXXX”

  api-server-host = “XXXX”

  webhookUrl = "https://hooks.slack.com/services/AAAAA”


  pipelines = [
  {
   name = "xhsecure-service-deploy-to-qa"
    stage = "deploy-to-qa"
    state = "failed|passed"
    channel = "sanket-test"
   webhookUrl = "https://hooks.slack.com/services/BBBBB”
  },
  {
    name = ".*"
    stage = ".*"
    state = "failed|passed"
    channel = "om-team-gocd"
  },
  {
    name = ".*"
    stage = ".*"
    state = "failed"
  }]
  # optional fields
  process-all-rules = "true"


}
```

**Comment by jdavisp3** [comment]
Oh, I see what you mean, right. I think I see where the bug might be.

**Comment by jdavisp3** [comment]
Oh, could you also make sure to add `#` to the channel names? It looks to me like the plugin expects them to start with `#`.

**Comment by jdavisp3** [comment]
BTW, I think you could work around this bug by specifying `webhookURL` for every rule.

**Comment by chakrapani123** [comment]
tried with both those changed settings..didnt work either

**Comment by ashwanthkumar** [comment]
Closing this issue for being non-active right now. If you're still facing the issues please re-open the issue.
